redhat-developer/service-binding-operator
# Immutable Binding Secrets

### Motivation

This PR:

1. Makes binding secrets immutable.
- Each time binding data is updated, a new binding secret is created with the name format `<name of ServiceBinding> + "-" + <hash of the binding data>`
- Example: `sb-1-72ddc0c540ab3a290e138726940591debf14c581` where `sb-1` is the name of ServiceBinding and `72ddc0c540ab3a290e138726940591debf14c581` is the hash of the binding data.

2. Fixes #630 

### Changes

- [x] Removes `changeTriggerEnv` which was used earlier to force redeployment of the application.
- [x] Removes SBR annotations from the application
- [x] Adds code to sort the secret keys and create data hash
- [x] Adds an acceptance test for the scenario https://github.com/redhat-developer/service-binding-operator/issues/630
- [x] Updates existing unit and acceptance test

### Steps to test this PR

1. Use one of the examples given [here](https://github.com/redhat-developer/service-binding-operator/tree/master/examples) to create a service, an application and a ServiceBinding. Ensure binding is successful by checking the status of the ServiceBinding.

2. Execute `oc get deployment <application_name> -o json | jq '.spec.template.spec.containers` and notice binding secret name is present

3. Update binding data in the service.

4. Again, Execute `oc get deployment <application_name> -o json | jq '.spec.template.spec.containers` and notice new secret name is present.

5. Create another ServiceBinding refering to the same application.

6. Check the application pods `oc get pods` and notice constant creation and termination of pods has stopped.

7. Also, check Service Binding Operator logs to see that indefinite reconciliation has stopped on applying two ServiceBindings referring to the same application.


### Acceptance Test

[1] To run the acceptance tests, execute `make test-acceptance`
